### PR TITLE
MVP

### PR DESCRIPTION
**Pull request by rmxr**
Put together a playable concept

### CODE REVIEW

File: .prettierrc.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (ViktorZhurbin) [inline .prettierrc.json]: usually, it is an empty `{}` config, so you may as well put it in package.json, like: `"prettier": {}`, and remove this file

File: package.json [hunk 2]
```diff
@@ -22,6 +22,7 @@
     "eslint": "^8.38.0",
     "eslint-plugin-react-hooks": "^4.6.0",
     "eslint-plugin-react-refresh": "^0.3.4",
+    "prettier": "2.8.8",
```
Comment (ViktorZhurbin) [inline package.json:32]: you need to set up npm command to actually use it, like `"prettier:fix": "prettier 'src/**/*.{ts,tsx,json}' --write"`. Then make sure to run `npm run prettier:fix` automatically at some point, like in a Github Action pipeline.

Or, my recommendation, use a pre-commit git hook - https://typicode.github.io/husky/getting-started.html. You can use it for both eslint and prettier. It will check you code when you commit it locally, making sure no poop sneaks in.

We can set up a Github Action as a second line of defence later - just for practice.

EDIT: same goes for `npm run lint`

Comment (ViktorZhurbin) [reply package.json:32]: Also, I recommend installing Eslint and Prettier extensions for VS code. Eslint will show you the errors, Prettier can format code On save

File: src/utils/types.ts
Comment (ViktorZhurbin) [inline src/utils/types.ts:1]: I suggest that types should live separately: in `types/` or `models/`. It's not utils

Comment (ViktorZhurbin) [reply src/utils/types.ts:1]: it's ok for now but if the app grows, Card and Field types should live in separate files

File: src/components/Card/Card.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import styles from "./Card.module.css";
+const back =
+  "https://images.unsplash.com/photo-1550537687-c91072c4792d?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=774&q=80";
```
Comment (ViktorZhurbin) [inline src/components/Card/Card.tsx]: Maybe download the image and keep it in `assets/` folder. This will save you from an awkward situation if it is removed or url gets changed

File: src/components/Card/Card.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  image: string;
+  index: number;
+  off: boolean;
+  flip: boolean;
+  clickHandler: (index: number, url: string) => void;
+}) {
+  return (
+    <div
+      onClick={() => {
+        clickHandler(index, image);
```
Comment (ViktorZhurbin) [inline src/components/Card/Card.tsx]: Looks like you don't need `index` in this file other than here. You can define handler in the parent component and only pass `clickHandler` prop to `Card`.

Usually, such props are named `onClick`

File: src/components/Card/Card.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      }}
+      className={`${styles.card} ${!flip ? styles.flipped : ""} ${
```
Comment (ViktorZhurbin) [inline src/components/Card/Card.tsx]: this sound illogical:  if `not flip` apply `styles.flipped`

File: src/components/Card/Card.module.css [hunk 6]
```diff
@@ -0,0 +1,57 @@
+.container {
```
Comment (ViktorZhurbin) [inline src/components/Card/Card.module.css]: this class is unused in Card?

File: src/components/Card/Card.module.css [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    top: 0;
+    left: 0;
+    height: 100%;
+    overflow: hidden;
+    border-radius: calc(var(--card-width)*0.1);
+}
+
+.front {
+    width: 100%;
+    z-index: 2;
```
Comment (ViktorZhurbin) [inline src/components/Card/Card.module.css]: why z-index is required?

File: src/components/Card/Card.module.css [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    transform: rotateY(0deg);
+}
+
+.back {
+    width: 100%;
```
Comment (ViktorZhurbin) [inline src/components/Card/Card.module.css]: `width` can be moved to common back and front styles

File: src/components/Card/Card.module.css [hunk 9]
```diff
@@ -0,0 +1,57 @@
+.container {
+    perspective: 1000px;
+}
+
+.card {
```
Comment (ViktorZhurbin) [inline src/components/Card/Card.module.css:1]: Just a thought: maybe add some `cursor: pointer`?

File: src/components/Field/Field.tsx
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx:1]: Field is a bit ambiguous name. Maybe PlayField same as the type interface?

File: src/components/Field/Field.tsx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+      };
+    });
+    setField(newField);
+  }, []);
+  React.useEffect(() => {
+    console.log(openCards);
+    setTimeout(() => {
+      handleCards();
+    }, "1000");
+  }, [openCards]);
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: Screenshot 2023-06-03 at 15 55 31

Eslint and TS would notify about warnings and errors. Pay attention to them

**Review by ViktorZhurbin — CHANGES_REQUESTED** [review]

File: src/utils/util.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+export function shuffleArray(array: string[]) {
+  for (let i = array.length - 1; i > 0; i--) {
+    const j = Math.floor(Math.random() * (i + 1));
+    [array[i], array[j]] = [array[j], array[i]];
```
Comment (ViktorZhurbin) [inline src/utils/util.ts]: Changing input array (mutation) is generally considered harmful. Your input array lives somewhere else and the order of items in it is changed on every function call. If you want to use it somewhere else, random order may give you unexpected results - and debugging will be difficult. Since this is a util function, it shouldn't rely on a particular use case where you don't care about the order.

Instead, either create a new variable from it at the start (`const array = [ ...inputArray ]`), or use non-mutating array methods that return a new array (like `.map` and `.reduce`). And return the new array at the end.

Comment (ViktorZhurbin) [inline src/utils/util.ts]: I'd also suggest to make it more readable. This Leetcode-style/algorythm-problem-solver js is hard to follow. Better naming would help others (and maybe even you) better understand what's going on.

Consider this:
```js
const lastIndex = array.length - 1;

for (let index = lastIndex; index > 0; index--) {
    const newIndex = Math.floor(Math.random() * array.length);
        
    [array[index], array[newIndex]] = [array[newIndex], array[index]];
}
```

File: src/components/Field/Field.tsx [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+  React.useEffect(() => {
+    const newField = images.map((image) => {
+      return {
+        url: image,
+        flip: false,
+        off: false,
+      };
+    });
+    setField(newField);
+  }, []);
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: you shuffle `images` on every render. Not sure that's intended. Maybe move this out of the component. Overall, it seems like you can extract this logic like:
```tsx
// assuming that shuffleArray returns a shuffled array
const initialImageUrls = shuffleArray([...initialImages, ...initialImages]);

const initialField: IPlayingFieldItem[] = initialImageUrls.map((url) => ({
    url,
    flip: false,
    off: false,
}));

// above code might as well become a `getInitialField` helper function that lives in src/Field/helpers.ts
// but make sure to call it outside component, like const initialField = getInitialField()

function Field() {
const [field, setField] = React.useState<IPlayingFieldItem[]>(initialField);
// you don't need initial useEffect now
...
}
```

File: src/components/Field/Field.tsx [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  React.useEffect(() => {
+    console.log(openCards);
+    setTimeout(() => {
+      handleCards();
+    }, "1000");
+  }, [openCards]);
+  function handleCards() {
+    if (openCards.length == 2) {
+      if (openCards[0].url === openCards[1].url) {
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: be consistent with equality checks. `===` is generally the best practice. `==` is tricky

File: src/components/Field/Field.tsx [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        };
+        newField[openCards[1].index] = {
+          ...field[openCards[1].index],
+          flip: false,
+        };
+        setField(newField);
+        setOpenCards([]);
+      }
+    }
+    console.log(openCards);
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: remove log? :)

File: src/components/Field/Field.tsx [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    return;
+  }
+  const handleClick = (index: number, url: string) => {
+    const newField = [...field];
+    newField[index] = { ...field[index], flip: true };
+    setField(newField);
+    if (openCards.length == 0 || openCards[0].index !== index) {
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: I'll be annoying about naming, sorry :))

```js
// now I can read var name instead of unwrapping openCards[0].index == index in my head
const isFirstOpenCard = openCards[0].index == index;

if (openCards.length == 0 || !isFirstOpenCard) { ... }
```

File: package.json (hunk 2 above)
Comment (rmxr) [reply package.json:32]: I set up ESLint and Prettier plugins in WebStorm. Will it suffice or I still have to do the pre-commit hook thing?

File: src/components/Field/Field.tsx [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+        setOpenCards([]);
+      }
+    }
+    console.log(openCards);
+    return;
+  }
+  const handleClick = (index: number, url: string) => {
+    const newField = [...field];
+    newField[index] = { ...field[index], flip: true };
+    setField(newField);
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: If card is already open, this would still call `setField` and cause re-render. I suggest to use early return if below code doesn't need to run:

```
const card = field[index];

if (card?.flip) {
  return;
}

// Now you don't need to check for openCards[0].index == index below?
```

File: src/components/Field/Field.tsx [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    if (openCards.length == 0 || openCards[0].index !== index) {
+      const newOpenCards = [...openCards];
+      newOpenCards.push({ index: index, url: url });
+      setOpenCards(newOpenCards);
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: `{ index: index, url: url }` can be just `{ index, url}` - using shorthand property names.

And all three lines can be reduced to `setOpenCards([...openCards, { index, url }]);`

File: src/components/Field/Field.tsx [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+    }
+  };
+
+  return (
+    <div className={styles.container}>
+      {field &&
+        field.map(({ url, flip, off }, index) => (
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: What are you checking by `field && field.map( ... )`?

File: src/components/Field/Field.tsx [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+          <Card
+            flip={flip}
+            off={off}
+            index={index}
+            key={index}
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: I'd use `url` as key. In this particular case it's not important, but generally it's safer to use some unique identifier for react key prop. If the order of items changes, index 2 would point to a different item, and react can't figure it out , you get bugs, and you may spend a lot of time debugging it

You can leave it as is, but I just want to let you know about a potential pitfall

**Review by ViktorZhurbin — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (ViktorZhurbin) [reply package.json:32]: Kind of, but there's a few points:
- you haven't yet formatted all the code - if I run `npx prettier --write .` it formats quite a few files.

For one-developer project it may be fine. But it's not reliable.
- if I commit to this repo and I don't have plugins installed, poorly formatted code will make it in to the code base
- if you get a new laptop, set up editor and forget to install plugins - you won't notice any warnings/errors

Also, I think it's worth learning how teams work on codebases - you won't always work alone - and code-formatting + linting are important parts of it these days

File: src/components/Field/Field.tsx [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+        const newField = [...field];
+        newField[openCards[0].index] = {
+          ...field[openCards[0].index],
+          flip: false,
+        };
+        newField[openCards[1].index] = {
+          ...field[openCards[1].index],
+          flip: false,
+        };
+        setField(newField);
```
Comment (ViktorZhurbin) [inline src/components/Field/Field.tsx]: What if you do `setField(initialField)`? You just want initial state here?

File: src/components/Field/Field.tsx (hunk 19 above)
Comment (ViktorZhurbin) [reply src/components/Field/Field.tsx]: urls are duplicated, so we can leave as is

File: src/utils/util.ts
Comment (ViktorZhurbin) [inline src/utils/util.ts:1]: I'd just rename the file to `array.ts` and only keep array helpers here

File: src/utils/util.ts [hunk 21]
```diff
@@ -0,0 +1,8 @@
+export function shuffleArray(arr: string[]) {
+  const array = [...arr];
+  for (let i = array.length - 1; i > 0; i--) {
+    const newIndex = Math.floor(Math.random() * array.length);
+    [array[i], array[newIndex]] = [array[newIndex], array[i]];
```
Comment (ViktorZhurbin) [inline src/utils/util.ts]: It's a knit-pick, but `i` and `newIndex` look like an inconsistent naming. Either name `i` > `index`, or `newIndex` > `newI`

File: src/components/PlayField/PlayField.tsx [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+  return {
+    url,
+    flip: false,
+    off: false,
+  };
+});
+
+function PlayField() {
+  const [openCards, setOpenCards] = React.useState<IOpenCardsItem[]>([]);
+  const [field, setField] = React.useState<IPlayingFieldItem[]>(initialField);
```
Comment (ViktorZhurbin) [inline src/components/PlayField/PlayField.tsx]: What if you name this `[cards, setCards]`? Will it make sense? Then `cards.map((card) => { ... }` would make more sense to me at least

File: src/components/PlayField/PlayField.tsx [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+  const handleCards = React.useCallback(() => {
```
Comment (ViktorZhurbin) [inline src/components/PlayField/PlayField.tsx]: `handleCards` doesn't explain the purpose of the function. What does is do exactly? Check result or something else? Then `handleCheckResult` or smething like that would be more self-explanatory

File: src/components/PlayField/PlayField.tsx [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+        };
+        newField[openCards[1].index] = {
+          ...field[openCards[1].index],
+          flip: false,
+        };
+        setField(newField);
+        setOpenCards([]);
+      }
+    }
+    return;
```
Comment (ViktorZhurbin) [inline src/components/PlayField/PlayField.tsx]: you don't need this `return` I think

File: src/components/PlayField/PlayField.tsx [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+  }, [field, openCards]);
+  React.useEffect(() => {
+    setTimeout(() => {
+      handleCards();
```
Comment (ViktorZhurbin) [inline src/components/PlayField/PlayField.tsx]: it seems more logical to move `if (openCards.length === 2) {` here
```js
if (openCards.length === 2) {
  handleCards();
}
```

This way you can clearly see that we want to check results only when two cards open

File: src/components/PlayField/PlayField.tsx [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+    }, 1000);
+  }, [handleCards, openCards]);
+
+  const handleClick = (index: number, url: string) => {
+    if (openCards.length === 2) {
```
Comment (ViktorZhurbin) [inline src/components/PlayField/PlayField.tsx]: `openCards.length === 2` is the final condition, used twice in the component. I suggest to assign it to a variable with a good name

File: src/components/PlayField/PlayField.tsx [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+      return;
+    }
+    const card = field[index];
+    if (card.flip) {
+      return;
+    }
+    const newField = [...field];
+    newField[index] = { ...field[index], flip: true };
+    setField(newField);
+    if (openCards.length < 2) {
```
Comment (ViktorZhurbin) [inline src/components/PlayField/PlayField.tsx]: maybe add a constant, like `const MAX_OPEN_CARDS = 2` or alike, and reuse it across the project. It's an important part of the game mechanics

File: src/components/Card/Card.module.css [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+  --card-width: 250px;
```
Comment (ViktorZhurbin) [inline src/components/Card/Card.module.css]: Is this the same value used in `PlayField` `grid-template-columns: repeat(4, 250px);`? If so, then it's better to define and use it there - it will also be available here.

File: src/constants/initialImages.ts [hunk 29]
```diff
@@ -0,0 +1,8 @@
+export const initialImages = [
```
Comment (ViktorZhurbin) [inline src/constants/initialImages.ts:1]: I'd suggest to download images and keep them as assets. Then you don't need to worry if url changes some day in the future or if the image is simply deleted

**Review by ViktorZhurbin — CHANGES_REQUESTED** [review]

File: src/components/PlayField/PlayField.tsx [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+  const handleClick = (index: number, url: string) => {
+    if (openCards.length === 2) {
+      return;
+    }
+    const card = field[index];
+    if (card.flip) {
+      return;
+    }
+    const newField = [...field];
+    newField[index] = { ...field[index], flip: true };
```
Comment (ViktorZhurbin) [inline src/components/PlayField/PlayField.tsx]: You can re-use card variable in `{ ...card, flip: true }`

File: src/components/PlayField/PlayField.tsx [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+          ...field[openCards[0].index],
+          flip: false,
+        };
+        newField[openCards[1].index] = {
+          ...field[openCards[1].index],
+          flip: false,
+        };
+        setField(newField);
+        setOpenCards([]);
+      }
```
Comment (ViktorZhurbin) [inline src/components/PlayField/PlayField.tsx]: I was trying to refactor this code and came up with this. Does it look better to you?

```js
      const [cardOne, cardTwo] = openCards;

      const newField = [...field];
      const newCardOne = newField[cardOne.index];
      const newCardTwo = newField[cardTwo.index];
      const isSuccess = cardOne.url === cardTwo.url;

      if (isSuccess) {
        newCardOne.off = true;
        newCardTwo.off = true;
      } else {
        newCardOne.flip = false;
        newCardTwo.flip = false;

      }

      setField(newField);
      setOpenCards([]);
```

File: src/constants/initialImages.ts (hunk 29 above)
Comment (rmxr) [reply src/constants/initialImages.ts:1]: I'd rather not do it for now for two reasons:
1. I want to keep the app lightweight.
2. As a possible feature, I want to implement the player's ability to change image sets

File: src/components/PlayField/PlayField.tsx (hunk 31 above)
Comment (rmxr) [reply src/components/PlayField/PlayField.tsx]: This one is beautiful

File: src/components/PlayField/PlayField.tsx (hunk 22 above)
Comment (ViktorZhurbin) [reply src/components/PlayField/PlayField.tsx]: and `initialField` > `initialCards` then?

File: src/components/PlayField/PlayField.tsx (hunk 26 above)
Comment (ViktorZhurbin) [reply src/components/PlayField/PlayField.tsx]: if you want to skip this one - reply here, so that I know for sure you saw this comment

**Review by ViktorZhurbin — APPROVED** [review]
🎉 🔥 Two minor comments left - up to you to resolve them
